onnxscript/ir/_core.py

```diff
@@ -1621,6 +1650,10 @@ def const_value(
         The value can be backed by different raw data types, such as numpy arrays.
         The only guarantee is that it conforms TensorProtocol.
         """
+        if self._const_value != None:
+            return self._const_value
+        # On-demand computation of constant value. Currently limited to outputs of Constant nodes.
+        self._compute_const_value()
```

Comment:
I am hesitant to include this because const_value is a user set field (not computed). Maybe we can have a new api called (`propagate_constants()`) or something like that on the value, to compute the constants?

Reply:
Looks like it's a bit tricky with some tradeoffs:
* The case where the above maybe tricky/unexpected is if the user explicitly sets this field to `None` (but it is the output of a Constant node). If the user sets the field to any non-None value, the behavior is still as expected. Setting the value to None is not expected or used right now, I think, especially for the output of a Constant node. We could check for it in the setter too, to disallow it.
* I think I can move the call to _compute_const_value to the Node(...) constructor. I think that may be a good choice.
* I prefer that users don't have to remember to call propagate() before looking up the constant-value.
* Most of the approaches still have some risk if nodes are explicitly modified. Eg., I could change the op_type from `Constant` to something else. I guess, for now, we can leave it as the user-responsibility to clear the constant-value if they make mutations that affect the const-value.

Reply:
The main concern I have for computing const_value is the potential overhead for accessing this field, since it is a property, we have conveyed an expectation of O(1) access. The same for the Node constructor - it would be good to stay lightweight imo. 

I wonder if there is a way to do it more elegantly? For example, would it be a good idea to allow something like `propagate()` to return a constant tensor (and set const_value as well)? This way when users expect computed values they will just call propagate() (or `compute()` or some properly named method) without having to do the two step access?

This way passes can still use the const_value field to store constants for their own usage, and when users need a computed value they can call this new method to obtain it.

Reply:
Just to clarify the tradeoffs above: calling `_compute_const_value` lazily (as in the current PR) is a bit more robust than calling it earlier (when the value is first created as the output of a specific node). Eg., it would work better if the user is allowed to create a node with given inputs and given number of outputs, and the user sets the domain/op_type etc. for the node subsequently. That's not an idiom we use or want to encourage, but it is doable. But I am not sure if it will make a difference in our current implementation.

Reply:
The overhead is still O(1). I don't think the cost significant enough to really matter. I think we can add a boolean field to ensure that this overhead happens only the first time it is called. (I did think about it, but didn't bother since it just eliminates a string comparison more or less, but it can be done).

We can change it to *not* be a property if that will help. But I am concerned about exposing two different APIs (const_value and propagate) which will increase the likelihood that somebody ends up using const_value instead of propagate, which will be a bigger concern.

Reply:
@justinchuby  I have removed this change. Now, it is the responsibility of any traversal/mutator to ensure that the const_value is set/updated as needed. Which is less than ideal, but hopefully it works fine for now, will need to try it on the entire benchmark.